File: swift/Examples/Deep/Deep.swift
import ArgumentParser
import HelpTree

struct Deep: ParsableCommand {
    static var configuration = CommandConfiguration(
        commandName: "deep",
        abstract: "A deeply nested CLI example (3 levels)",
        subcommands: [Server.self, Client.self]
    )

    @Flag(name: .customLong("help-tree"), help: "Print a recursive command map derived from framework metadata")
    var helpTree: Bool = false

    @Option(name: [.customShort("L"), .customLong("tree-depth")], help: "Limit --help-tree recursion depth")
    var treeDepth: Int?

    @Option(name: [.customShort("I"), .customLong("tree-ignore")], help: "Exclude subtrees/commands from --help-tree output")
    var treeIgnore: [String] = []

    @Flag(name: [.customShort("a"), .customLong("tree-all")], help: "Include hidden subcommands in --help-tree output")
    var treeAll: Bool = false

    @Option(name: .customLong("tree-output"), help: "Output format (text or json)")
    var treeOutput: String?

    @Option(name: .customLong("tree-style"), help: "Tree text styling mode (rich or plain)")
    var treeStyle: String?

    @Option(name: .customLong("tree-color"), help: "Tree color mode (auto, always, never)")
    var treeColor: String?
}

struct Server: ParsableCommand {
    static var configuration = CommandConfiguration(
        commandName: "server",
        abstract: "Server management",
        subcommands: [Config.self, Db.self]
    )
}

struct Config: ParsableCommand {
    static var configuration = CommandConfiguration(
        commandName: "config",
        abstract: "Configuration commands",
        subcommands: [Get.self, Set.self, Reload.self]
    )
}

struct Db: ParsableCommand {
    static var configuration = CommandConfiguration(
        commandName: "db",
        abstract: "Database commands",
        subcommands: [Migrate.self, Seed.self, Backup.self]
    )
}

struct Client: ParsableCommand {
    static var configuration = CommandConfiguration(
        commandName: "client",
        abstract: "Client operations",
        subcommands: [Auth.self, Request.self]
    )
}

struct Auth: ParsableCommand {
    static var configuration = CommandConfiguration(
        commandName: "auth",
        abstract: "Authentication commands",
        subcommands: [Login.self, Logout.self, Whoami.self]
    )
}

struct Request: ParsableCommand {
    static var configuration = CommandConfiguration(
        commandName: "request",
        abstract: "HTTP request commands",
        subcommands: [Get.self, Post.self]
    )
}

struct Get: ParsableCommand {
    static var configuration = CommandConfiguration(commandName: "get", abstract: "Get a config value")
    @Argument(help: "Key or path")
    var key: String
}

struct Set: ParsableCommand {
    static var configuration = CommandConfiguration(commandName: "set", abstract: "Set a config value")
    @Argument(help: "Key")
    var key: String
    @Argument(help: "Value")
    var value: String
}

struct Reload: ParsableCommand {
    static var configuration = CommandConfiguration(commandName: "reload", abstract: "Reload configuration")
}

struct Migrate: ParsableCommand {
    static var configuration = CommandConfiguration(commandName: "migrate", abstract: "Run migrations")
}

struct Seed: ParsableCommand {
    static var configuration = CommandConfiguration(commandName: "seed", abstract: "Seed the database")
}

struct Backup: ParsableCommand {
    static var configuration = CommandConfiguration(commandName: "backup", abstract: "Backup the database")
}

struct Login: ParsableCommand {
    static var configuration = CommandConfiguration(commandName: "login", abstract: "Log in")
}

struct Logout: ParsableCommand {
    static var configuration = CommandConfiguration(commandName: "logout", abstract: "Log out")
}

struct Whoami: ParsableCommand {
    static var configuration = CommandConfiguration(commandName: "whoami", abstract: "Show current user")
}

struct Post: ParsableCommand {
    static var configuration = CommandConfiguration(commandName: "post", abstract: "Send a POST request")
    @Argument(help: "Path")
    var path: String
}

@main
struct DeepEntry {
    static func main() {
        let args = Array(CommandLine.arguments.dropFirst())
        if let invocation = HelpTree.parseInvocation(args) {
            HelpTree.run(for: Deep.self, invocation: invocation)
        } else {
            Deep.main()
        }
    }
}

File: c/examples/deep.c
#include <stdio.h>
#include <string.h>
#include "../src/help_tree.h"

static const ht_option_t verbose_opt = {"verbose", "", "--verbose", "Verbose output", false, false, "", false};

static const ht_command_t config_get = {
    "get", "Get a config value", &verbose_opt, 1,
    (const ht_argument_t[]){ {"KEY", "Config key", true, false} }, 1, NULL, 0, false
};
static const ht_command_t config_set = {
    "set", "Set a config value", &verbose_opt, 1,
    (const ht_argument_t[]){ {"KEY", "Config key", true, false}, {"VALUE", "Config value", true, false} }, 2, NULL, 0, false
};
static const ht_command_t config_reload = {
    "reload", "Reload configuration", &verbose_opt, 1, NULL, 0, NULL, 0, false
};
static const ht_command_t config_cmd = {
    "config", "Configuration commands", &verbose_opt, 1, NULL, 0,
    (const ht_command_t[]){config_get, config_set, config_reload}, 3, false
};

static const ht_command_t db_migrate = {"migrate", "Run migrations", NULL, 0, NULL, 0, NULL, 0, false};
static const ht_command_t db_seed = {"seed", "Seed the database", NULL, 0, NULL, 0, NULL, 0, false};
static const ht_command_t db_backup = {"backup", "Backup the database", NULL, 0, NULL, 0, NULL, 0, false};
static const ht_command_t db = {
    "db", "Database commands", NULL, 0, NULL, 0,
    (const ht_command_t[]){db_migrate, db_seed, db_backup}, 3, false
};

static const ht_command_t server = {
    "server", "Server management", &verbose_opt, 1, NULL, 0,
    (const ht_command_t[]){config_cmd, db}, 2, false
};

static const ht_command_t auth_login = {"login", "Log in", NULL, 0, NULL, 0, NULL, 0, false};
static const ht_command_t auth_logout = {"logout", "Log out", NULL, 0, NULL, 0, NULL, 0, false};
static const ht_command_t auth_whoami = {"whoami", "Show current user", NULL, 0, NULL, 0, NULL, 0, false};
static const ht_command_t auth = {
    "auth", "Authentication commands", NULL, 0, NULL, 0,
    (const ht_command_t[]){auth_login, auth_logout, auth_whoami}, 3, false
};

static const ht_command_t req_get = {
    "get", "Send a GET request", &verbose_opt, 1,
    (const ht_argument_t[]){ {"PATH", "Request path", true, false} }, 1, NULL, 0, false
};
static const ht_command_t req_post = {
    "post", "Send a POST request", &verbose_opt, 1,
    (const ht_argument_t[]){ {"PATH", "Request path", true, false} }, 1, NULL, 0, false
};
static const ht_command_t request = {
    "request", "HTTP request commands", &verbose_opt, 1, NULL, 0,
    (const ht_command_t[]){req_get, req_post}, 2, false
};

static const ht_command_t client = {
    "client", "Client operations", &verbose_opt, 1, NULL, 0,
    (const ht_command_t[]){auth, request}, 2, false
};

static const ht_option_t root_opts[] = {
    {"help-tree", "", "--help-tree", "Print a recursive command map derived from framework metadata", false, false, "", false},
    {"tree-depth", "-L", "--tree-depth", "Limit --help-tree recursion depth (Unix tree -L style)", false, true, "", false},
    {"tree-ignore", "-I", "--tree-ignore", "Exclude subtrees/commands from --help-tree output (repeatable)", false, true, "", false},
    {"tree-all", "-a", "--tree-all", "Include hidden subcommands in --help-tree output", false, false, "", false},
    {"tree-output", "", "--tree-output", "Output format (text or json)", false, true, "", false},
    {"tree-style", "", "--tree-style", "Tree text styling mode (rich or plain)", false, true, "", false},
    {"tree-color", "", "--tree-color", "Tree color mode (auto, always, never)", false, true, "", false},
    {"verbose", "", "--verbose", "Verbose output", false, false, "", false},
};

static const ht_command_t root = {
    "deep", "A deeply nested CLI example (3 levels)",
    root_opts, sizeof(root_opts)/sizeof(root_opts[0]), NULL, 0,
    (const ht_command_t[]){server, client}, 2, false
};

int main(int argc, char **argv) {
    ht_invocation_t *inv = ht_parse_invocation(argc - 1, argv + 1);
    if (inv) {
        ht_config_file_t *cfg = ht_load_config("examples/help-tree.json");
        ht_apply_config(&inv->opts, cfg);
        ht_run_for_tree(&root, &inv->opts, inv->path, inv->path_count);
        ht_free_config(cfg);
        ht_free_invocation(inv);
        return 0;
    }
    printf("Run with --help-tree to see the command tree.\n");
    return 0;
}
